features/dd-sdk-android-trace/src/main/java/com/datadog/trace/core/propagation/W3CHttpCodec.java

```diff
@@ -28,347 +28,348 @@
-        // Override Datadog sampling priority with W3C one
-        this.propagationTags.updateTraceSamplingPriority(
-            contextPriority, SamplingMechanism.EXTERNAL_OVERRIDE);
-      } else {
-        // Use more detailed Datadog sampling priority in context
-        this.samplingPriority = ptagsPriority;
-      }
-      // Use the origin
-      this.origin = this.propagationTags.getOrigin();
-      // Ensure TraceId high-order bits match
-      this.propagationTags.updateTraceIdHighOrderBits(this.traceId.toHighOrderLong());
-    }
+        /**
+         * We need to delay handling of the W3C headers until the end since tracestate headers should be
+         * concatenated before processed, and only if there is a single valid traceparent header present
+         */
+        @Override
+        protected TagContext build() {
+            // If there is neither a traceparent nor a tracestate header, then ignore this
+            if (traceparentHeader == null && tracestateHeader == null) {
+                onlyTagContext();
+            }
+            if (valid && fullContext) {
+                try {
+                    if (traceparentHeader == null && tracestateHeader != null) {
+                        throw new IllegalStateException(
+                                "Found no traceparent header but tracestate header '" + tracestateHeader + "'");
+                    }
+                    // Now we know that we have at least a traceparent header
+                    parseTraceParentHeader(traceparentHeader);
+                    parseTraceStateHeader(tracestateHeader);
+                } catch (RuntimeException e) {
+                    onlyTagContext();
+                    log.debug("Exception when building context", e);
+                }
+            }
+            return super.build();
+        }
 
-    private static String trim(String input) {
-      if (input == null) {
-        return "";
-      }
-      final int last = input.length() - 1;
-      if (last == 0) {
-        return input;
-      }
-      int start;
-      for (start = 0; start <= last; start++) {
-        char c = input.charAt(start);
-        if (c != '\t' && c != ' ') {
-          break;
+        void parseTraceParentHeader(String tp) {
+            int length = tp == null ? 0 : tp.length();
+            if (length < TRACE_PARENT_LENGTH) {
+                throw new IllegalStateException("The length of traceparent '" + tp + "' is too short");
+            }
+            long version = LongStringUtils.parseUnsignedLongHex(tp, 0, 2, true);
+            if (version == 255) {
+                throw new IllegalStateException("Illegal version number " + tp.substring(0, 2));
+            } else if (version == 0 && length > TRACE_PARENT_LENGTH) {
+                throw new IllegalStateException("The length of traceparent '" + tp + "' is too long");
+            }
+            DDTraceId traceId = DD128bTraceId.fromHex(tp, TRACE_PARENT_TID_START, 32, true);
```

Comment:
it is going to come later in the next PRs according with Ganesh but Id id not do anything here, this class was modified just because of the format change.